Tasks/AzureMonitorAlerts/azurerestutility.ts

```diff
@@ -0,0 +1,265 @@
+import * as tl from "vsts-task-lib/task";
+import * as Q from "q";
+import * as querystring from "querystring";
+
+import { 
+    HttpClient, 
+    HttpClientResponse 
+} from "typed-rest-client/HttpClient";
+
+import { 
+    RestClient, 
+    IRequestOptions, 
+    IRestResponse 
+} from "typed-rest-client/RestClient";
+
+import * as Util from "./utility";
+import { 
+    IAzureMetricAlertRequestBody,
+    IAzureMetricAlertRule
+} from "./interfaces";
+
+var httpObj = new HttpClient(tl.getVariable("AZURE_HTTP_USER_AGENT"));
+var restObj = new RestClient(tl.getVariable("AZURE_HTTP_USER_AGENT"));
+
+var defaultAuthUrl = 'https://login.windows.net/';
+
+function getAuthorizationToken(endPoint): Q.Promise<string> {
+
+    let deferred: Q.Deferred<string> = Q.defer<string>();
+    let envAuthUrl = (endPoint.envAuthUrl) ? (endPoint.envAuthUrl) : defaultAuthUrl;
+    let authorityUrl = envAuthUrl + endPoint.tenantID + "/oauth2/token/";
+    let requestData = querystring.stringify({
+        resource: endPoint.activeDirectoryResourceId,
+        client_id: endPoint.servicePrincipalClientID,
+        grant_type: "client_credentials",
+        client_secret: endPoint.servicePrincipalKey
+    });
+
+    let requestHeader = {
+        "Content-Type": "application/x-www-form-urlencoded; charset=utf-8"
+    }
+
+    tl.debug('Requesting for Auth Token: ' + authorityUrl);
+    httpObj.post(authorityUrl, requestData, requestHeader)
+        .then(async (response: HttpClientResponse) => {
+            if (response.message.statusCode == 200) {
+                let contents: string = await response.readBody();
+                if(!!contents) {
+                    deferred.resolve(JSON.parse(contents).access_token);
+                }
+            }
+            else {
+                deferred.reject(tl.loc('CouldnotfetchaccesstokenforAzureStatusCode', response.message.statusCode, response.message.statusMessage));
+            }
+        }, (error) => {
+            deferred.reject(error);
+        });
+
+    return deferred.promise;
+}
+
+export async function getAzureRmResourceDetails(
+    endpoint, 
+    resourceUri: string) : Promise<any> {
+    
+    let deferred: Q.Deferred<any> = Q.defer<any>();
+    let accessToken: string = await getAuthorizationToken(endpoint);
```

Comment:
the response to the getauthtoken call contains the expiration time for the accesstoken. Should we use that to change the logic in a way, such that rest call is made only in case that the accesstoken is expired. @kmkumaran , I remember i did a change for something similar, but at that time, we didn't do it because we did not want to break anything and also because at that time the getauthtoken calls were not many (this was sometime during the time of app service deploy version 2) . Now we call the getaccesstoken a whole lot of time. But since this a new task. we should give the idea a try. If it works, it would help us to reduce the no. of getauthtoken calls. significantly

Reply:
+1

Reply:
👍